packages/google_maps_flutter/example/lib/place_marker.dart

```diff
@@ -25,142 +25,188 @@ class PlaceMarkerBody extends StatefulWidget {
   State<StatefulWidget> createState() => PlaceMarkerBodyState();
 }
 
+typedef Marker MarkerUpdateAction(Marker marker);
+
 class PlaceMarkerBodyState extends State<PlaceMarkerBody> {
   PlaceMarkerBodyState();
 
   static final LatLng center = const LatLng(-33.86711, 151.1947171);
 
   GoogleMapController controller;
-  int _markerCount = 0;
-  Marker _selectedMarker;
+  Map<MarkerId, Marker> markers = <MarkerId, Marker>{};
+  MarkerId selectedMarker;
+  int _markerIdCounter = 1;
 
   void _onMapCreated(GoogleMapController controller) {
     this.controller = controller;
-    controller.onMarkerTapped.add(_onMarkerTapped);
   }
 
   @override
   void dispose() {
-    controller?.onMarkerTapped?.remove(_onMarkerTapped);
     super.dispose();
   }
 
-  void _onMarkerTapped(Marker marker) {
-    if (_selectedMarker != null) {
-      _updateSelectedMarker(
-        const MarkerOptions(icon: BitmapDescriptor.defaultMarker),
-      );
-    }
+  void _updateSelectedMarker(MarkerUpdateAction update) {
     setState(() {
-      _selectedMarker = marker;
+      if (markers.containsKey(selectedMarker)) {
+        final Marker newMarker = update(markers[selectedMarker]);
+        markers[selectedMarker] = newMarker;
+      }
     });
-    _updateSelectedMarker(
-      MarkerOptions(
-        icon: BitmapDescriptor.defaultMarkerWithHue(
-          BitmapDescriptor.hueGreen,
-        ),
-      ),
-    );
-  }
-
-  void _updateSelectedMarker(MarkerOptions changes) {
-    controller.updateMarker(_selectedMarker, changes);
   }
 
   void _add() {
-    controller.addMarker(MarkerOptions(
+    final int markerCount = markers.length;
+
+    if (markerCount == 12) {
+      return;
+    }
+
+    final String markerIdVal = 'marker_id_$_markerIdCounter';
+    _markerIdCounter++;
+    final MarkerId markerId = MarkerId(markerIdVal);
+
+    void _onMarkerTapped() {
```

Comment:
I think it'll simplify things if we avoid this closure.
We can have a member `_onMarkerTapped(MarkerId markerId);` and below do:

```dart
    final Marker marker = Marker(
      // ...
      onTap: () { _onMarkerTapped(markerId); },
    );
```

This way the logic one sees in the _add method is purely about adding a marker.